lib/react_client.dart

```diff
@@ -734,8 +751,56 @@ class ReactJsComponentFactoryProxy extends ReactComponentFactoryProxy {
   }
 }
 
+class ReactDartFunctionComponentFactoryProxy extends ReactComponentFactoryProxy
+    with JsBackedMapsComponentFactoryBuilder {
+  /// The ReactJS function used as the type for all [ReactElement]s built by
+  /// this factory.
+  final Function reactFunction;
+
+  /// The name of this function.
+  final String displayName;
+
+  /// The JS component factory used by this factory to build [ReactElement]s.
+  final ReactJsComponentFactory reactComponentFactory;
+
+  ReactDartFunctionComponentFactoryProxy(JsFunctionComponent reactFunction, [String displayName])
+      : this.reactFunction = reactFunction,
+        this.displayName = displayName,
+        this.reactComponentFactory = React.createFactory(reactFunction);
+
+  get type => reactFunction;
+}
+
+/// Creates a function component from the given [dartFunctionComponent] that can be used with React.
+JsFunctionComponent _wrapFunctionComponent(DartFunctionComponent dartFunctionComponent, {String displayName}) {
+  assert(() {
+    displayName ??= getProperty(dartFunctionComponent, 'name');
```

Comment:
This could use a similar comment to the one on existing code https://github.com/cleandart/react-dart/blob/a3b562304c2f73f95ee6253ef1f0f62141e622c5/lib/react.dart#L603-L607 to make it clear why we're doing this.

Reply:
after re-thinking about this... we chose to not use runtimeType due to dart2js de-optimization, but here its just a property getter, that should be pretty fast, do we even need the assert? ive confirmed it still works with -r

Reply:
> However the dart2js deoptimization concerns are less about runtime speed at the call site and more about increased size of code output. I'll try to find a reference for this and post back.

I believe that the output code from a getProperty is pretty minor, in fact i think it ends up becoming like `self.dartFunctionComponent.name` ? or something like that hence my thought on removing the assert.

Reply:
> Actually, it looks like `toString` might be an exception: [dart-lang/linter#1475](https://github.com/dart-lang/linter/issues/1475) so maybe we don't need the assert after all!

also that sounds super promising! do you mean we could potentially remove the assert from component2 as well since we use the toString?